### PR TITLE
Changes for handling recursive NHG entries

### PR DESCRIPTION
**Pull request by nkelapur**
Changes to handle recursive NHG entries

| Repo | Title |    PR   |
|--|--|   --|
| sonic-swss | Recursive nexthop group enhancement  | https://github.com/sonic-net/sonic-swss/pull/3105   |

### CODE REVIEW

**Comment by nkelapur** [comment]
This change is part of https://github.com/sonic-net/SONiC/issues/1594

**Comment by adyeung** [comment]
@nakano-omw please help review the change in nexthop group

**Comment by prsunny** [comment]
@nkelapur , @utpalkantpintoo, is this discussed in Sonic community meeting?

**Comment by adyeung** [comment]
@prsunny the proposal was presented and reviewed in the Routing WG meetings, this is a HLD submission based on the WG discussion

**Comment by prsunny** [comment]
> @prsunny the proposal was presented and reviewed in the Routing WG meetings, this is a HLD submission based on the WG discussion

ack

File: doc/pic/hld_fpmsyncd.md [hunk 1]
```diff
@@ -99,6 +102,8 @@ For multipath route, the `RTM_NEWROUTE` is sent with a list of gateway and inter
 
 This `Fpmsyncd extension` will modify `fpmsyncd` to handle  `RTM_NEWNEXTHOP` and `RTM_DELNEXTHOP` as below.
 
+This featured introduces new orchagent `NhgOrch` to handle `RTM_NEWNEXTHOP` events and update `NEXTHOP_GROUP_TABLE` entries accordingly
```
[the remark this answers is not in the document]
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md]: **Done**

File: doc/pic/hld_fpmsyncd.md [hunk 2]
```diff
@@ -248,6 +262,32 @@ Therefore, even after this enhancement, table entries will be created for `ROUTE
 ##### Figure: Example of ASIC_DB entry
 ![fig3](images_fpmsyncd/fig3.svg)
 
+#### Orchestration Agent Changes to handle NEXTHOP_GROUP
+
+A new orchestration agent `NhgOrch` will be written to handle the new NEXTHOP_GROUP_TABLE in APP_DB. For adding or updating an entry in the NEXTHOP_GROUP_TABLE, programming will depend on whether the group is configured with one or multiple next hops or is a recursive nexthop group i.e. it contains one or more nexthop group.
```
[the remark this answers is not in the document]
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md]: Done

File: doc/pic/hld_fpmsyncd.md [hunk 3]
```diff
@@ -475,6 +515,46 @@ Sample of CONFIG DB snippet given below:
...
+  Stores a list of groups of one or more next hops
+  Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; 
+                    arbitrary string identifying the next hop 
+                    group, as determined by the programming 
+                    application.
+    nexthop       = *prefix,           ; IP addresses separated
+                     “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated 
+                     by “,” (zero indicates no interface)
```
[the remark this answers is not in the document]
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:534]: This table is new in this HLD. Added "existing" and "new" fields labels for the "ROUTE_TABLE"

File: doc/pic/hld_fpmsyncd.md [hunk 4]
```diff
@@ -475,6 +515,46 @@ Sample of CONFIG DB snippet given below:
...
+                    by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole
+                   (or null0)
+    nexthop_group = NEXT_HOP_GROUP_TABLE:key ; index within the
+                    NEXT_HOP_GROUP_TABLE, optionally used instead
+                    of nexthop and intf fields
+```
+
 
 ### Warmboot and Fastboot Design Impact
```
[the remark this answers is not in the document]
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:559]: Done

**Comment by adyeung** [comment]
@arlakshm the HLD has been reviewed in the Routing WG, let us know if you have further comment, team is looking to close next week

File: doc/pic/hld_fpmsyncd.md [hunk 5]
```diff
@@ -484,6 +564,7 @@ Mention whether this feature/enhancement has got any requirements/dependencies/i
 
 - When the feature is disabled, there should be no impact to Warmboot and Fastboot.
 - When the feature is enabled, there will be no warmboot nor fastboot support.
```
Comment (hasan-brcm) [inline doc/pic/hld_fpmsyncd.md:566]: Fast boot can be supported. Only WB will have limitations due to reconciliation of nhids.

File: doc/pic/hld_fpmsyncd.md [hunk 6]
```diff
@@ -475,6 +515,46 @@ Sample of CONFIG DB snippet given below:
         }
     },
 ```
+#### APPL DB Enhancements  
+
+This feature adds a new NEXT_HOP_GROUP_TABLE, to store next hop group information to be used by one or more routes. Recursive/ecmp routes are represented by referencing a list of nexthop group ids in the nexthop_group field. When this field is used, other fields will not be present.
```
Comment (hasan-brcm) [inline doc/pic/hld_fpmsyncd.md:520]: "adds a new .."  => "extends the existing .."

Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md]: Done

File: doc/pic/hld_fpmsyncd.md (hunk 5 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md]: Done

**Review by hasan-brcm — APPROVED** [review]

File: doc/pic/hld_fpmsyncd.md [hunk 7]
```diff
@@ -475,6 +515,46 @@ Sample of CONFIG DB snippet given below:
...
+### NEXT_HOP_GROUP_TABLE
+  Stores a list of groups of one or more next hops
+  Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; 
+                    arbitrary string identifying the next hop 
+                    group, as determined by the programming 
+                    application.
+    nexthop       = *prefix,           ; IP addresses separated
+                     “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated
```
Comment (venkatmahalingam) [inline doc/pic/hld_fpmsyncd.md:533]: Why just PORT_TABLE? how about portchannel, vlan and sub-interfaces..etc

File: doc/pic/hld_fpmsyncd.md [hunk 8]
```diff
@@ -475,6 +515,46 @@ Sample of CONFIG DB snippet given below:
...
+  Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; 
+                    arbitrary string identifying the next hop 
+                    group, as determined by the programming 
+                    application.
+    nexthop       = *prefix,           ; IP addresses separated
+                     “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated 
+                     by “,” (zero indicates no interface)
+    nexthop_group = NEXT_HOP_GROUP_TABLE:key, ; index within the
```
Comment (venkatmahalingam) [inline doc/pic/hld_fpmsyncd.md:535]: can you add an example for the recursive routes?

File: doc/pic/hld_fpmsyncd.md [hunk 9]
```diff
@@ -230,14 +235,23 @@ admin@sonic:~$ sonic-db-cli APPL_DB HGETALL ROUTE_TABLE:10.1.1.4
...
 
-- When receiving `RTM_NEWNEXTHOP` events on sequence 1, 2 and 3, `fpmsyncd` will save the information in an internal list to be used when necessary.
-- When `fpmsyncd` receive the `RTM_NEWROUTE` on sequence 4, the process will write the NextHop Group with ID 118 to the `NEXTHOP_GROUP_TABLE` using the information of gateway and interface from the NextHop Group events with IDs 116 and 117.
-- Then `fpmsyncd` will create a new route entry to `ROUTE_TABLE` with a `nexthop_group` field with value `ID118`.
-- When `fpmsyncd` receives the last `RTM_NEWROUTE` on sequence 5, the process will create a new route entry (but no NextHop Group entry) in `ROUTE_TABLE` with `nexthop_group` field with value `ID118`. (Note: This NextHop Group entry was created when the `fpmsyncd` received the event sequence 4.)
+- When receiving `RTM_NEWNEXTHOP` events on sequence 1 and 2, `fpmsyncd` will write the information in `NEXTHOP_GROUP_TABLE` with NextHop Group `ID125` and `ID126` and with information of gateway and interface respectively. For sequence 3  multi path NextHop Group `ID127` the information in `NEXTHOP_GROUP_TABLE` will contain the list of IDs 125 and 126
```
Comment (venkatmahalingam) [inline doc/pic/hld_fpmsyncd.md:249]: What changes are expected towards kernel programming? please capture those details if not captured in a separate document.

File: doc/pic/hld_fpmsyncd.md [hunk 10]
```diff
@@ -248,6 +262,32 @@ Therefore, even after this enhancement, table entries will be created for `ROUTE
 ##### Figure: Example of ASIC_DB entry
 ![fig3](images_fpmsyncd/fig3.svg)
 
+#### Orchestration Agent Changes to handle NEXTHOP_GROUP
+
+Orchestration agent `NhgOrch` will be enhanced to handle the new NEXTHOP_GROUP_TABLE in APP_DB. For adding or updating an entry in the NEXTHOP_GROUP_TABLE, programming will depend on whether the group is configured with one or multiple next hops or is a recursive nexthop group i.e. it contains one or more nexthop group.
```
Comment (Gokulnath-Raja) [inline doc/pic/hld_fpmsyncd.md:267]: Could see code related to handling of Loopback interface and NHID added as CPU port, please capture here if required.

File: doc/pic/hld_fpmsyncd.md [hunk 11]
```diff
@@ -248,6 +262,32 @@ Therefore, even after this enhancement, table entries will be created for `ROUTE
 ##### Figure: Example of ASIC_DB entry
 ![fig3](images_fpmsyncd/fig3.svg)
 
+#### Orchestration Agent Changes to handle NEXTHOP_GROUP
+
```
Comment (Gokulnath-Raja) [inline doc/pic/hld_fpmsyncd.md:266]: Please capture the case NH as Interface?

File: doc/pic/hld_fpmsyncd.md (hunk 7 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:533]: There is no change in this DB Table. This is as is from https://github.com/sonic-net/SONiC/blob/master/doc/ip/next_hop_group_hld.md

File: doc/pic/hld_fpmsyncd.md (hunk 9 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:249]: No new changes are required from frr/kernel side for this feature.  When NHG is enabled in FRR, it already programs routes to kernel using reference to the NHID.  This implementation in fact make the FPM programming similar to how the routes are being programmed in kernel already

File: doc/pic/hld_fpmsyncd.md (hunk 7 above)
Comment (venkatmahalingam) [reply doc/pic/hld_fpmsyncd.md:533]: Do we allow NH interfaces other than PORT_TABLE?

File: doc/pic/hld_fpmsyncd.md (hunk 8 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:535]: The DB entry for recursive NHG is same as ECMP NHGs. In that for recursive NHG may contain only one NH and hence the NHA_GROUP will contain only one NHID

File: doc/pic/hld_fpmsyncd.md (hunk 11 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:266]: "NH as Interface" will be same as any other NHG.  The application ( zebra) will create a NHID for the NHG with NH as interface and the route will use this NHID as reference to the NH

File: doc/pic/hld_fpmsyncd.md (hunk 7 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:533]: Yes, as I understand all possible existing NH interface would be allowed

File: doc/pic/hld_fpmsyncd.md [hunk 12]
```diff
@@ -230,14 +235,23 @@ admin@sonic:~$ sonic-db-cli APPL_DB HGETALL ROUTE_TABLE:10.1.1.4
...
+- When `fpmsyncd` receive the `RTM_NEWROUTE` on sequence 4, the process will create a new route entry to `ROUTE_TABLE` with a `nexthop_group` field with value `ID127`.
+- When `fpmsyncd` receives the next `RTM_NEWROUTE` on sequence 5, the process will create a new route entry (but no NextHop Group entry) in `ROUTE_TABLE` with `nexthop_group` field with value `ID127`. (Note: This NextHop Group entry was created when the `fpmsyncd` received the event sequence 4.)
+- When receiving `RTM_NEWNEXTHOP` events on sequence 6 `fpmsyncd` will write the information in `NEXTHOP_GROUP_TABLE` with NextHop Group `ID128`
+- When receiving `RTM_NEWNEXTHOP` events on sequence 7 `fpmsyncd` will update the information in `NEXTHOP_GROUP_TABLE` with NextHop Group `ID127` and the information will now contain list of IDs 125 and 128
+
 
 #### Example of entries in ASIC_DB
```
[the remark this answers is not in the document]
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:257]: Added diagram to show the update ASIC_DB entry

File: doc/pic/hld_fpmsyncd.md (hunk 7 above)
Comment (venkatmahalingam) [reply doc/pic/hld_fpmsyncd.md:537]: Just mentioning PORT_TABLE alone the schema is weird, please fix it.

Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md]: Changed to "INTF_TABLE" so that all types on interfaces are covered

File: doc/pic/hld_fpmsyncd.md (hunk 10 above)
Comment (nkelapur) [reply doc/pic/hld_fpmsyncd.md:271]: This change has been removed from the code PR now

**Review by Gokulnath-Raja — APPROVED** [review]

**Review by venkatmahalingam — APPROVED** [review]

**Review by adyeung — APPROVED** [review]

**Pull request merged**